pkg/authorization/authorizer/authorizer.go

```diff
@@ -0,0 +1,513 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return policyBindingList.Items, nil
+}
+
+// getRoleBindings provides a point for easy caching
+func (a *openshiftAuthorizer) getRoleBindings(namespace string) ([]authorizationapi.RoleBinding, error) {
+	policyBindings, err := a.getPolicyBindings(namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make([]authorizationapi.RoleBinding, 0, len(policyBindings))
+	for _, policyBinding := range policyBindings {
+		for _, value := range policyBinding.RoleBindings {
+			ret = append(ret, value)
+		}
+	}
+
+	return ret, nil
+}
+
+// getRole
+func (a *openshiftAuthorizer) getRole(roleBinding authorizationapi.RoleBinding) (*authorizationapi.Role, error) {
+	roleNamespace := roleBinding.RoleRef.Namespace
+	roleName := roleBinding.RoleRef.Name
+
+	rolePolicy, err := a.getPolicy(roleNamespace)
+	if err != nil {
+		return nil, err
+	}
+
+	role, exists := rolePolicy.Roles[roleName]
+	if !exists {
+		return nil, fmt.Errorf("role %#v not found", roleBinding.RoleRef)
+	}
+
+	return &role, nil
+}
+
+func (a *openshiftAuthorizer) getEffectivePolicyRules(namespace string, user authenticationapi.UserInfo) ([]authorizationapi.PolicyRule, error) {
+	roleBindings, err := a.getRoleBindings(namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	effectiveRules := make([]authorizationapi.PolicyRule, 0, len(roleBindings))
+	for _, roleBinding := range roleBindings {
+		role, err := a.getRole(roleBinding)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, curr := range role.Rules {
+			if doesApplyToUser(roleBinding.UserNames, roleBinding.GroupNames, user) {
+				effectiveRules = append(effectiveRules, curr)
+			}
+		}
+	}
+
+	return effectiveRules, nil
+}
+
+func (a *openshiftAuthorizer) Authorize(passedAttributes AuthorizationAttributes) (bool, string, error) {
+	// fmt.Printf("#### checking %#v\n", passedAttributes)
+
+	attributes, ok := passedAttributes.(openshiftAuthorizationAttributes)
+	if !ok {
+		return false, "", fmt.Errorf("attributes are not of expected type: %#v", attributes)
+	}
+
+	globalAllowed, globalDenied, err := a.authorizeWithNamespaceRules(a.masterAuthorizationNamespace, attributes)
+	if err != nil {
+		return false, "", err
+	}
+	if len(globalDenied) > 0 {
+		return false, globalDenied, err
+	}
+	if len(globalAllowed) > 0 {
+		return true, globalAllowed, err
+	}
+
+	if len(attributes.GetNamespace()) != 0 {
+		namespaceAllowed, namespaceDenied, err := a.authorizeWithNamespaceRules(attributes.GetNamespace(), attributes)
+		if err != nil {
+			return false, "", err
+		}
+		if len(namespaceDenied) > 0 {
+			return false, namespaceDenied, err
+		}
+		if len(namespaceAllowed) > 0 {
+			return true, namespaceAllowed, err
+		}
+	}
+
+	return false, "denied by default", nil
+}
+
+func (a *openshiftAuthorizer) authorizeWithNamespaceRules(namespace string, passedAttributes AuthorizationAttributes) (string, string, error) {
```

Comment:
triggering off of reason lengths feels odd. can this return `result, reason, err`, where result is `allow`, `deny`, or FileNotFound?